presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
             checkState(rawInputs != null && blockBuilders == null);
             rawInputs.ensureCapacity(rawInputsLength);
 
+            // Perform dedup for distinct aggregates
+            if (delegate instanceof DistinctingGroupedAccumulator) {
+                // Prepend the groupId block to the input page
+                Page withGroup = page.prependColumn(groupIdsBlock);
+
+                // Compute a mask for the distinct rows (including the group id)
+                checkState(markDistinctHash.isPresent());
+                Work<Block> work = markDistinctHash.get().markDistinctRows(withGroup);
+                checkState(work.process());
+                Block distinctMask = work.getResult();
+
+                // Filter out duplicate rows
+                Page dedupPage = filter(withGroup, distinctMask);
+
+                // Update GroupByIdBlock from distinctPage
+                groupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+                // Remove the groupId block from distinct Page
+                page = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```

Comment:
Now my question is which step here actually removes the non-distinct rows from memory?

Reply:
OK, I wasn't sure if `getPositions` on Block actually always creates a copy or a view of the block. Looking at the implementation in say `RunLengthEncodedBlock` it makes no sense to me because it discards the positions and just returns another wrapper. `DictionaryBlock` also seems to play similar games.

The default implementation creates a DictionaryBlock which I think is just a wrapper around the source block?

Reply:
I think the key difference is that the filtered stuff is normally fed to the underlying aggregation. So the reduction happens there and most of this is concerned with identifying what to feed to it. But here I think we just store the page away because we won't feed it to the aggregation until much later.

Reply:
> Depending on how MarkDistinctHash works is it going to store these values twice?

MarkDistinctHash will store each distinct row only once. It internally uses GroupByHash which has all the logic to detect duplicates. When duplicate row comes in, it computes the hash for this duplicate row and if this hash has been seen before, then it performs an actual check of values if they are same or not. If equality check passes, then it means that it is actually a duplicate row. If equality check does not passes, then it means that there was a hash collision and it is handled accordingly. So, basically, each distinct row is stored only once.

> If there are multiple distinct aggregations on the same column with no filter or the same filter then we would also like to share the memory used to track distinct values.

Yes, that would be good to have. But, I am not targeting this in this PR. Currently, each aggregate has its own MarkDistinctHash that is used to perform deduplication. Maybe, this can be done as an optimization in followup PR.

Reply:
I probably put out too much noise so the signal got lost here.

`filter` doesn't actually filter. It returns a `Page` that appears to be a view of the original `Page`'s blocks. This is retaining all non-distinct values for the distincted columns. Please correct me if I am wrong, but all the `getPositions` implementations I looked at create views of the original `Block`.

`MarkDistinctHash` copies the column values being distincted so we are storing the values twice. Once in `MarkDistinctHash` and once in the pages that we fed to mark distinct hash, but also retained. We don't need to do that. We could only retain `MarkDistinctHash` and then extract the values stored in it.

Reply:
> filter doesn't actually filter. It returns a Page that appears to be a view of the original Page's blocks. This is retaining all non-distinct values for the distincted columns. Please correct me if I am wrong, but all the getPositions implementations I looked at create views of the original Block.

This is correct. `filter` calls `getPositions` which creates a view over the existing block. But, when we prepare pages for spill, we only spill filtered values and during unspill,  create a page with only distinct rows. We can actually call `copyPositions` instead which performs the actual copy and returns a new page which only contains distinct value.  This will incur some cpu cost due to copy. How expensive is copy operation? If we go with copy approach, we will save on memory buffer as only distinct values will be kept and buffered.


> MarkDistinctHash copies the column values being distincted so we are storing the values twice. Once in MarkDistinctHash and once in the pages that we fed to mark distinct hash, but also retained. We don't need to do that. We could only retain MarkDistinctHash and then extract the values stored in it.

This is actually a good observation. We can actually get values directly from MarkDistinctHash but MarkDistinctHash only stores the value for hash channels. It does not store values for maskChannel which is also required in DistinctingGroupedAccumulator in some cases. Its possible to modify markDistinctHash to start storing maskChannel values as well. Let me give it a deeper thought tomorrow.

Reply:
I think the issue here is that turning on spilling with this implementation causes queries to use more memory thus forcing them to spill and possibly run out of spill space.

To roll out spilling for distinct aggregation for all queries we need to be at least as memory efficient as the non-spilling version. There is no justification for a regression and on the wider workload it really will cause problems.

To target specific queries where it helps and doesn't hurt then sure we can use this approach, but it is pretty wrong to store non-distinct values in memory. That isn't the right behavior for aggregation.